### ISSUE TITLE
Bayesian analysis for calibration of input perturbation

### ISSUE THREAD

**Issue by SorooshMani-NOAA** [noaa-ocs-modeling/EnsemblePerturbation#88]

**Comment by SorooshMani-NOAA** [comment]
@WPringle based on the discussion we had today, we only need to run best track for a couple of storms and with a couple of different ensemble sizes with Korobov sampling right?

**Comment by WPringle** [comment]
@SorooshMani-NOAA because it is best track we should change how we do the ensemble perturbation. I think we need to work out that strategy first, then we do ensemble then check that the ensemble results encompass the observations.

This paper is relevant which also look at best track:
Sochala, P., Chen, C., Dawson, C., & Iskandarani, M. (2020). A polynomial chaos framework for probabilistic predictions of storm surge events. Computational Geosciences, 24, 109–128. https://doi.org/10.1007/s10596-019-09898-5

See section 4.2. In there they treat the errors as uniform random variables with some amplification factor which we can tune. 
Check out Figure 5 and 6.

I think we should follow their methodology for doing the track and intensity perturbations.

**Comment by WPringle** [comment]
@SorooshMani-NOAA You can see they also perturb manning's roughness too.

**Comment by SorooshMani-NOAA** [comment]
Just for the sake of completeness of this discussion:
We discussed how to generate ensembles and eventually decided to just run ensembles of different storms with different sizes for now. This is due to the limitation we had on using the available resources in terms of time.

Later we'll explore other parameter perturbations.

**Comment by SorooshMani-NOAA** [comment]
@WPringle have you tried running KL-NN code on the sample data? During our meeting (and also in the slides) it says the data is time dependent. But looking at the `ydata.txt` and the code that was shared with us, it seems that it's using data that is of the shape `ensemble x mesh-node`. Do you know if the plots in the slides where generated using the same code or not?

I haven't tried running it yet. I was just reviewing the slides and the code.

**Comment by WPringle** [comment]
I did run the code using the sample data and the plots on the slides are the result of that execution.

I wouldn't worry about whether it is "x" or "t", it is somewhat arbitrary and methodology can handle spatiotemporal in theory. But I am pretty sure "x" is referring to a time dimension here based on how it looks

**Comment by SorooshMani-NOAA** [comment]
I see, thank you

**Comment by SorooshMani-NOAA** [comment]
@WPringle I forgot to also ask, have you tried it on an ensemble too? Was there any issue I need to be aware of if I want to run this on combine maxelev ensemble dataset?

**Comment by WPringle** [comment]
No I did not. Just be aware of the dimensions of the data, number of parameters, number of nodes, and number of ensembles and how that fits into that code. It should be able to run through to the part where it requires the observations

**Comment by SorooshMani-NOAA** [comment]
@WPringle I'm testing the KLNN (without MCMC part) on the same ensemble I used with your KLPC. I actually subsetted the maxelev dataset using your subsetting part of the KLPC script (from [`examples/karhunenloeve_polynomialchaos/klpc_wetonly.py`](https://github.com/noaa-ocs-modeling/EnsemblePerturbation/blob/main/examples/karhunenloeve_polynomialchaos/klpc_wetonly.py)). Although the dataset now fits into memory, still the `KLE.build` takes a lot of time. Would it be OK if we replace the KL part of Khachik's script with the same PCA that you were using in the following code? https://github.com/noaa-ocs-modeling/EnsemblePerturbation/blob/33141b8c4586c22a6e3923d6f3fff7ed95c4b9b3/ensembleperturbation/uncertainty_quantification/karhunen_loeve_expansion.py#L39

Out of curiosity, is there any reason behind using KLE/PCA for reducing dimensions? Why not use other methods?

**Comment by WPringle** [comment]
Yes, please use the PCA method we use in `EnsemblePerturbation`

What other methods are there? For the KLPC we have an analytical solution to convert the KL fit surrogate back to the full-dimension surrogate, which is at least one advantage I can think of...

**Comment by SorooshMani-NOAA** [comment]
> Yes, please use the PCA method we use in EnsemblePerturbation

OK thanks!


> What other methods are there?

I was talking in general, I know there are a bunch of dimension reduction techniques which might be better than PCA based certain ways/applications; so I was wondering if there's any specific reason. e.g.: https://towardsdatascience.com/11-dimensionality-reduction-techniques-you-should-know-in-2021-dcb9500d388b


> ... to convert the KL fit surrogate back to the full-dimension surrogate...

I see, I didn't really think about this part. It would still be interesting to see what other methods have to offer if we get the chance to explore

**Comment by WPringle** [comment]
yeah that's interesting! For sure explore them if you want to, we can have a meeting with Khachik sometime to discuss this, he might be able to say why we would prefer one type of method over another.

**Comment by SorooshMani-NOAA** [comment]
Copying the email for future reference:
> As I was going over the KL-NN-MCMC code thinking about how to use USGS data within the model, or later how to use the temporal data, I decided to focus more on the PCA part of the code and read more about it. I stumbled upon this not so new paper:
> 
> https://www.researchgate.net/publication/263138615_Principal_Component_Analysis_on_Spatial_Data_An_Overview
> 
> This is an overview of PCA use for spatial data. For the standard PCA that we are using, it defines different modes of running, and neither of them incorporates all time, location and attributes at the same time.
> 
> Based on my understanding, in our case node index is "location", each ensemble is an "attribute", and the elevation output time in the time series is "time" (for later). We could also consider both ensemble-member and the timeseries as "time", but I'm not sure if that makes a lot of sense. So right now for max-elev analysis we are actually in Q-mode, but I don't know which mode we should use for (later) time series analysis.
> 
> Another question that I have is how to see if we can find a way to "train" KLNN on all the nodes, but use USGS data on ~270 HWM locations for MCMC tuning. The alternative obvious approach is to train KLNN on the station locations and then do MCMC on the same station locations (noting that since we don't have a lot of computational time, we interpolate field outputs to the HWM locations for now!).

**Comment by SorooshMani-NOAA** [comment]
I tried using the `sklearn` PCA and the KLNN part goes through successfully now. I downloaded USGS data and selected all the locations that are within the polygon of the mesh. Since the stations were not available at the time of simulation, I'm interpolating results to the station locations for the analysis. Now I'm trying to fix `nan` data and/or dimension mismatches to run the full KLNNMCMC.

**Comment by WPringle** [comment]
@SorooshMani-NOAA Could you please share the code you have so far for the results we looked at yesterday with Khachik?

**Comment by SorooshMani-NOAA** [comment]
@WPringle I uploaded the code on a private repo for now. I just added you and Saeed as collaborator. Note that the code must run on an already subsetted dataset. I tried to get it to work with minimum changes possible.

**Comment by SorooshMani-NOAA** [comment]
This is the plot I get for the MCMC results. If I discard the first half and pick every 10th:
![image](https://user-images.githubusercontent.com/77082694/202289183-72753011-998f-4257-9209-7149608d6ed3.png)

Each plot is for one of the parameters values in Y axis vs X axis, the sample index. It seems that `gamma` in the code plays a major role in determining how white-noise like the parameters are; this if for `gamma=0.01`

**Comment by SorooshMani-NOAA** [comment]
I also increased the 2000 epochs to 20,000 and it definitely improved the NN:
![image](https://user-images.githubusercontent.com/77082694/202291072-82473c4a-3079-4308-94ee-b6f4ce746658.png)
and
![image](https://user-images.githubusercontent.com/77082694/202291166-c6e6ef47-f396-42cd-8978-3075a6df7449.png)

**Comment by SorooshMani-NOAA** [comment]
@WPringle if I want to train on the whole model but only MCMC on the USGS HWM stations, what should the non-station dimension value be when transforming observations using PCA? Should it be 0?

**Comment by WPringle** [comment]
> @WPringle if I want to train on the whole model but only MCMC on the USGS HWM stations, what should the non-station dimension value be when transforming observations using PCA? Should it be 0?

@SorooshMani-NOAA That's a good question. I suppose 0 makes sense because in the matrix multiplication zeros will add nothing. Or if it doesn't work then we need to probably do the MCMC on the full result, rather than the dimensionally reduced result. First let's try zero and see...

**Comment by SorooshMani-NOAA** [comment]
> Just to be clear, you are asking about what dummy value to put in obs_data at non-station locations right?

@WPringle, sorry for late reply. Yes that's what I meant. I'll try and see how it goes

**Comment by SorooshMani-NOAA** [comment]
On a separate note, I plotted the MCMC results vs obs as well as model vs obs and it doesn't really look good. Note that I've already cut off HWM locations that are higher than 2 meter or less that 1 meter outside the min/max of HWM station resutls in the model and still it doesn't look good. I think I need to actually rerun the simulations:
![image](https://user-images.githubusercontent.com/77082694/202780049-f9331336-d681-422b-aa83-a6c38cdac1d0.png)
where each thin bar is an station and the limits indicate max/min results from all the 30. and 
![image](https://user-images.githubusercontent.com/77082694/202780278-3c013022-a05e-4494-ae56-7dbbf46c25f8.png)

**Comment by SorooshMani-NOAA** [comment]
This is a diagnostic plot for best track model vs obs
![bt_vs_obs_schism_n_adcirc](https://user-images.githubusercontent.com/77082694/203430538-070dce2b-ac57-4dd9-a40b-ef5e2516983a.png)
I couldn't find a recent ADCIRC run, but this is a Florence sim using ADCIRC from Zach's older files. It seems that the ADCIRC setup would result in better results in this specific case. I need to also check my other runs and if the best track for all of them is not better, maybe I need to use a smaller mesh for the run.

**Comment by WPringle** [comment]
What's the reason for the different number of points? Maybe the mesh region? I think it would help to plot the same points

**Comment by SorooshMani-NOAA** [comment]
> I think it would help to plot the same points

Yeah, that'd make more sense. As you said the difference is due to `nan` values I get during interpolation which is related to the meshed region.

**Comment by SorooshMani-NOAA** [comment]
This is the result, I think I need to look at my workflow ensemble results to see if it's any better. If not I might just need to improve the mesh and rerun.
![bt_vs_obs_schism_n_adcirc](https://user-images.githubusercontent.com/77082694/203444446-fb5ffb1c-63c7-4680-a8fe-bdae1962c9d3.png)

btw, I forgot to include this one. It is the same fit plot, but without low quality hwm data:
![fit](https://user-images.githubusercontent.com/77082694/203445998-da2ca256-b48e-4ef4-92d1-522dce9c3ed6.png)

And this is the ensemble model results vs observation for high quality hwm data:
![model_vs_obs](https://user-images.githubusercontent.com/77082694/203446200-af9b3849-1622-429d-aa90-e5f0bc539bc9.png)

**Comment by SorooshMani-NOAA** [comment]
Using the full model + interpolated data on stations to train and then only stations for MCMC with 0 in other nodes:

![loss_history_log](https://user-images.githubusercontent.com/77082694/203604737-f2ff7ec6-cd38-469b-9307-b4cf25faf755.png)
![model_surr](https://user-images.githubusercontent.com/77082694/203604791-a0da62fb-0169-4006-86dc-cd13e1372a22.png)

It actually looks worse!
![fit](https://user-images.githubusercontent.com/77082694/203606161-a26e5671-94ba-4d8e-ab48-6b4d2053f05d.png)

![model_vs_obs](https://user-images.githubusercontent.com/77082694/203606135-158c522b-3977-4079-80fe-f53a3ba19caa.png)

**Comment by SorooshMani-NOAA** [comment]
The results of the two tasks in the previous comment were not very promising, I don't see any points in sharing the plots. One interesting point I noticed is that the ensemble from the workflow seem to have elevation biased compared to COOPS observations. I'm investigating it by plotting time series. However the strange part is that the test set-up was not different from the workflow setup other than the number of members & also the mesh (I'm not sure about the mesh), which shouldn't affect the best track, so how can a difference in best track results be explained?!

**Comment by WPringle** [comment]
@SorooshMani-NOAA About the NN size. Actually it should be mean of the total size of input and output. So input is size ndim*neig and output is size 1xneig, which means you can set the neuron size to be = neig(ndim+1)/2

**Comment by SorooshMani-NOAA** [comment]
All of the ensemble we ran during September 2022 used a version of SCHISM where parametric wind doesn't work correctly (commit `5282a27`). This means all of the ensembles we currently have are of no use and we have to run a new set!

I'm adding NWM and 48-hour before landfall perturbation, and we know that paramwind is fixed; so hopefully we'll get good quality results this time!

**Comment by SorooshMani-NOAA** [comment]
There's still some issues with NWM in the ensemble set-up. I'll just run a couple of ensemble with the fixed parametric wind to have preliminary results for ensemble analysis

**Comment by SorooshMani-NOAA** [comment]
For a new ensemble run for Florence this is what I get (observations from high quality coastal HWM):

**Observations vs best track**
![bt_vs_obs](https://user-images.githubusercontent.com/77082694/209158085-292be762-f31c-4d9b-91c4-c9b9dee47a22.png)

**Ensemble vs observation**
![model_vs_obs](https://user-images.githubusercontent.com/77082694/209171136-ff54f03c-5e1a-44d9-ab76-b937e700f45c.png)

**Surrogate vs Model**
![model_surr](https://user-images.githubusercontent.com/77082694/209171029-c8147487-b80e-42be-ad36-d1be87b210be.png)

**Convergence of NN**
![loss_history_log](https://user-images.githubusercontent.com/77082694/209158214-ab533482-9755-413c-aaaa-708da300b26c.png)

**Observation and Errorbars of MCMC**
![fit](https://user-images.githubusercontent.com/77082694/209170889-689303ef-4346-4b44-bf82-821286432874.png)

**Comment by SorooshMani-NOAA** [comment]
The result is not good for sure, but this time there seems to be a method to this madness! I mean there seems to be a bias involved here in the MCMC errorbar results. Still not sure how to interpret it, what do you think @WPringle ? Also please let me know how to share the `.nc` files. The combined for the 50 ensemble (42-43 went through) is around `300 GB`, the subset is ~`20GB`

**Comment by WPringle** [comment]
I guess the bias mostly comes from the fact that model results are generally under predicted. But it's a bit weird we would think the MAP crosses should be closest to the observations, instead they are lower down on the error bars.
Could you add the MCMC error bars to the Ensemble vs observation plot? So we can compare original ensemble to MCMC.

**Comment by WPringle** [comment]
@SorooshMani-NOAA I think we can use this function to transfer Gaussian to uniform https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.ndtr.html#scipy.special.ndtr

**Comment by WPringle** [comment]
@SorooshMani-NOAA I put my jupyter notebook and some modifications to `mcmc.py` and `utils.py` for using PCs on the globus Ensemble research collection under `KLPC_MCMC_William`

**Comment by WPringle** [comment]
Also check out https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html for trying to figure out best hyperparameters for the ANN model (used by this paper: https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2022JD037617)

**Comment by SorooshMani-NOAA** [comment]
@WPringle I think we missed one point during our discussion about Physical vs KL space:

When using Polynomial Chaos and we're in physical space, it's clear what is the input to the physical surrogate, but for the NN case, even if MCMC takes care of transforming back to physical, still we don't know what to pass to the NN in the first place. 

NN only accepts KL space input, so we still run into the same issue of transforming data from observation-points-only to KL space, to then pass it to NN. So we should find a way to "transform" the neural network to accept physical inputs. Does that make sense?

**Comment by WPringle** [comment]
No I think it is the same because the “input” to the surrogate is just the value of the set of parameters while the length of the surrogate is either the number of eigenmodes or number of physical nodes depending on if in KL space of physics space. Of course we need to also provide the observed data and the corresponding indices in the physical surrogate that map to the data.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by SorooshMani-NOAA** [comment]
> “input” to the surrogate is just the value of the set of parameters

Oh I see my mistake ... thanks! I'll just need to `inverse_transform` the KL space predictions from NN and then get the ones on obs location for MCMC to use!

##

Apart from that, I'm not sure if I use cross validation term correctly. Based on the description here https://scikit-learn.org/stable/modules/grid_search.html, this term is used for hyperparameter tuning (e.g. number of hidden layers or neurons in case of NN) and not model parameter selection. What we're doing really is looking at train/test for tuning parameters.

In the case of PC model, the existing code has:
```python
cross_validator = LeaveOneOut()
regression_model = ElasticNetCV(
    fit_intercept=False,
    cv=cross_validator,
    l1_ratio=0.5,
    selection='random',
    random_state=666,
)
```
based on https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNetCV.html my understanding is that cross validation is actually **not** used here, since `l1_ratio` is not a list of possible values.

**Comment by WPringle** [comment]
> In the case of PC model, the existing code has:
> 
> ```python
> cross_validator = LeaveOneOut()
> regression_model = ElasticNetCV(
>     fit_intercept=False,
>     cv=cross_validator,
>     l1_ratio=0.5,
>     selection='random',
>     random_state=666,
> )
> ```
> 
> based on https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNetCV.html my understanding is that cross validation is actually **not** used here, since `l1_ratio` is not a list of possible values.

Yep so it still is doing the CV for the `alpha` parameter in the cost function just not the  `l1_ratio` but yes that could also be added as a list.

**Comment by SorooshMani-NOAA** [comment]
> doing the CV for the alpha

Oh, yeah I missed it. In any case, what I'm trying to say is that unless we want to really find a good # of neurons or layers, CV approach might not make much sense for NN and we can do just the internal validation it already has for choosing parameters

**Comment by WPringle** [comment]
@SorooshMani-NOAA hmm I don't know about that. Maybe those CV functions aren't appropriate but I think it's important to do CV so we reduce the over training problem.

**Comment by SorooshMani-NOAA** [comment]
There seems to be other methods to avoid overfitting, e.g.:
https://www.kdnuggets.com/2019/12/5-techniques-prevent-overfitting-neural-networks.html
In any case, this is the results by using `val` input but without any further change to the NN code:
![loss_history_log](https://user-images.githubusercontent.com/77082694/216147042-1701cef7-bbce-49f0-a9c6-d38903623f05.png)
The best validation loss happens early in the epochs, so doing more and more epochs doesn't really help based on this graph.

And this is the observation vs ensemble (with dry node extrapolation):
![model_vs_obs](https://user-images.githubusercontent.com/77082694/216147840-6c5ee479-a69d-4ad9-8781-a3a19b6ac1ff.png)


Also the results from SCHISM ensemble also improved by using physical-space-surrogate:
![fit](https://user-images.githubusercontent.com/77082694/216146892-dcb3c5f8-2897-44af-ae13-ac19230bdb3f.png)

**Comment by WPringle** [comment]
@SorooshMani-NOAA See these scripts and best_track and advisory files here for Florence:
[florence_tracks_scripts.zip](https://github.com/noaa-ocs-modeling/EnsemblePerturbation/files/10795839/florence_tracks_scripts.zip)

The `plot_nhc_adv.py` uses the `tropycal` package and `get_advisory.py` uses `StormEvents`, but not sure if works properly for getting advisory. But it's supposed to be able to.

**Comment by WPringle** [comment]
[Sochala et al. - 2020 - A polynomial chaos framework for probabilistic predictions of storm surge events.pdf](https://github.com/noaa-ocs-modeling/EnsemblePerturbation/files/10911733/Sochala.et.al.-.2020.-.A.po
[Plumlee et al. - 2021 - High-fidelity hurricane surge forecasting using em.pdf](https://github.com/noaa-ocs-modeling/EnsemblePerturbation/files/10911737/Plumlee.et.al.-.2021.-.High-fidelity.hurricane.surge.forecasting.using.em.pdf)
lynomial.chaos.framework.for.probabilistic.predictions.of.storm.surge.events.pdf)

**Comment by WPringle** [comment]
[Sochala et al. - 2020 - A polynomial chaos framework for probabilistic predictions of storm surge events.pdf](https://github.com/noaa-ocs-modeling/EnsemblePerturbation/files/10911746/Sochala.et.al.-.2020.-.A.polynomial.chaos.framework.for.probabilistic.predictions.of.storm.surge.events.pdf)

**Comment by SorooshMani-NOAA** [comment]
@WPringle when I download data for all the past storms from `stormevents` it looks like there are only two `advisory_numbers`, `01` and `03`, which I think is different from that actual advisory ID that is based on the forecast start time. This comes from the ATCF track files. Based on this, here's how I'd like to get the 48hr forecast data for past storms using `stormevents`:

1. First get all the `OFCL` tracks for the storm using `VortexTrack.from_storm_name` by passing `advisories=['OFCL']` and `file_deck='a'`
2. Next get the underlying dataframe using either `.data` or `.unfiltered_data`
3. Find the intersection of the acquired dataframe with the shape of US (from `NaturalEarth_Lowres`)
4. For each `track_start_date` (eqv. to date from ATCF), get the **first** `datetime` (eqv. to date from ATCF + forecast hour) that the track point falls **on land**.
5. In the result of above operation get the difference of `datetime` and `track_start_date`
6. Get the last entry whose time diff is equal or immediately greater than 2 days to be the "48-hour forecast"
7. Use the track with the start time found in step 6 for perturbing.

What do you think of this approach?

**Comment by WPringle** [comment]
That sounds good to me @SorooshMani-NOAA. What do you mean by only two advisory_numbers?

**Comment by SorooshMani-NOAA** [comment]
If I get the tracks using:
```python
tracks = stormevents.nhc.track.VortexTrack.from_storm_name('florence', 2018, file_deck='a')
```
then this would be the result of `tracks.data`
```
      basin storm_number            datetime advisory_number advisory  ... isowave_radius_for_NWQ  isowave_radius_for_SWQ  extra_values                    geometry    track_start_time
0        AL           06 2018-08-29 06:00:00              01     CARQ  ...                    NaN                     NaN          <NA>  POINT (-10.90000 12.90000) 2018-08-30 06:00:00
1        AL           06 2018-08-29 12:00:00              01     CARQ  ...                    NaN                     NaN          <NA>  POINT (-12.30000 12.90000) 2018-08-30 06:00:00
2        AL           06 2018-08-29 18:00:00              01     CARQ  ...                    NaN                     NaN          <NA>  POINT (-13.80000 12.90000) 2018-08-30 06:00:00
3        AL           06 2018-08-30 00:00:00              01     CARQ  ...                    NaN                     NaN          <NA>  POINT (-15.40000 12.90000) 2018-08-30 06:00:00
4        AL           06 2018-08-30 06:00:00              01     CARQ  ...                    NaN                     NaN          <NA>  POINT (-17.00000 12.90000) 2018-08-30 06:00:00
...     ...          ...                 ...             ...      ...  ...                    ...                     ...           ...                         ...                 ...
10229    AL           06 2018-09-18 12:00:00              01     CARQ  ...                    NaN                     NaN          <NA>  POINT (-73.30000 42.20000) 2018-09-18 12:00:00
10230    AL           06 2018-09-18 12:00:00              01     CARQ  ...                    NaN                     NaN          <NA>  POINT (-73.30000 42.20000) 2018-09-18 12:00:00
10231    AL           06 2018-09-18 12:00:00              03     OFCL  ...                    NaN                     NaN          <NA>  POINT (-73.30000 42.20000) 2018-09-18 12:00:00
10232    AL           06 2018-09-18 15:00:00              03     OFCL  ...                    NaN                     NaN          <NA>  POINT (-71.90000 42.60000) 2018-09-18 12:00:00
10233    AL           06 2018-09-19 00:00:00              03     OFCL  ...                    NaN                     NaN          <NA>  POINT (-67.70000 43.50000) 2018-09-18 12:00:00
```
There's an `advisory_number` in the columns. Before looking at the values I thought this is just a number counting advisories as they are issued, but this seems to mean something different.

**Comment by WPringle** [comment]
yeah maybe that is being parsed incorrectly or the column in ALCF meant something else

**Comment by WPringle** [comment]
@SorooshMani-NOAA This looks very promising, nice job!
The problem you are encountering is associated with the fact that the advisories only provide the pressures at the 3-hr forecast and I think the Rmax also needs to be taken from initial time!

In my methodology the initial Rmax, maybe this is from CARQ, is assumed for all forecast times in the advisory.
The 0-hr pressure is also taken from the initial CARQ value.
The >3hr forecast pressures are computed as a function of Vmax by keeping Holland B constant. i.e., calculate Holland B for the 3-hr forecast and use that to get the pressures for all the forecast times in the future.

**Comment by WPringle** [comment]
@SorooshMani-NOAA This is one of big reasons why the Rmax methodology sucks because we although have a lot of information of Vmax and r34, r50, r64 we have no estimate of Rmax in the future. So better off to use r34, r50, r64 in a GAHM implementation

**Comment by SorooshMani-NOAA** [comment]
@WPringle, thanks. Do you have any suggestion for how to resolve this issue based on what we get above? I looked at the CARQ for the same `track_start_time` and the pressure is `0` there too.

**Comment by WPringle** [comment]
Do you get Rmax in the CARQ?

I think I remember now that actually for the pressure at the 0-hr I used the pressure from the 3-hr in the previous advisory (3 hours prior). If that is too complicated then we can fill in 0-hr using same methodology as the forecast times.

**Comment by SorooshMani-NOAA** [comment]
@WPringle I believe that should be what's happening. I'm just using `StormEvents` and I know that the correction code referred to above is being triggered. I believe based on the current `StormEvents` logic the pressure is indeed taken from 0-hr (the first entry of CARQ for the same track start time). 

I can manually modify some of the values in my workflow script to try different things., e.g. I can try setting it to 3-hr forecast pressure. Do you want us to change the logic in the `StormEvents` repo as well if it works when I try here?

**Comment by SorooshMani-NOAA** [comment]
Interesting ... I just noticed this in my dataframe, looking only at certain columns:
```
8909     CARQ            -24 2018-09-13 06:00:00 2018-09-12 06:00:00               0.0
8910     CARQ            -18 2018-09-13 06:00:00 2018-09-12 12:00:00               0.0
8911     CARQ            -12 2018-09-13 06:00:00 2018-09-12 18:00:00               0.0
8912     CARQ             -6 2018-09-13 06:00:00 2018-09-13 00:00:00               0.0
8913     CARQ              0 2018-09-13 06:00:00 2018-09-13 06:00:00             956.0
8914     CARQ              0 2018-09-13 06:00:00 2018-09-13 06:00:00             956.0
8915     CARQ              0 2018-09-13 06:00:00 2018-09-13 06:00:00             956.0                                                                                                                    9020     OFCL              0 2018-09-13 06:00:00 2018-09-13 06:00:00               0.0                                                                                                                    9021     OFCL              0 2018-09-13 06:00:00 2018-09-13 06:00:00               0.0                                                                                                                    9022     OFCL              0 2018-09-13 06:00:00 2018-09-13 06:00:00               0.0                                                                                                                    9023     OFCL              3 2018-09-13 06:00:00 2018-09-13 09:00:00             956.0                                                                                                                    9024     OFCL              3 2018-09-13 06:00:00 2018-09-13 09:00:00             956.0                        
```
I think the logic in StormEvents gets the first **entry** meaning `-24` hour. It needs to be changed to take the `hour 0` forecast instead ...

**Comment by WPringle** [comment]
@SorooshMani-NOAA yes!

**Comment by SorooshMani-NOAA** [comment]
@WPringle I just realized that in the original unprocessed ATCF files, the pressure fields are not `NA` or empty, they are just `0`. That's why `StormEvents` doesn't actually fix them, since it only updates the fields that are NA! I'll create a ticket on `StormEvents` side to keep track of updating this issue.
